qdrant-landing/content/articles/immutable.md

```diff
@@ -0,0 +1,135 @@
+```
+
+Vector Search engines, like Qdrant, have to deal with a large variety of data structures, but all of them have one thing in common: 
+if we could make them immutable, it could significantly help us to improve performance and optimize memory usage.
+
+## What immutability can improve exactly?
+
+Large part of immutable advantage comes from the fact, that we know the exact data we need to put into the structure even before we start building it.
+Simplest example is a sorted array: we would know exactly how many elements we have to put into the array, so we can allocate exact amount of memory once.
+
+More complex data structures might require additional statistics to be collected before the structure is built.
+Qdrant-related example of this is Scalar Quantization: on order to select proper quantization levels, we have to know the distribution of the data.
+
+(Image with quatiles here)
+
+Computing this distribution is requires knowing all the data in advance, but once we have it, applying scalar quantization is a simple operation.
+
+Let's take a look at non-exhaustive list of data structures and potential improvements we can get from making them immutable:
+
+|Function| Mutable Data Structure | Immutable Alternative | Potential improvements |
+|----|------|------|------------------------|
+| Read by index | Arrray | Fixed chunk of memory | Allocate exact amount of memory |
+| Vector Storage | Array or Arrays | Memory-mapped file | Offload data to disk |
+| Read sorted ranges| B-Tree | Sorted Array | Store all data close, avoid cache misses |
+| Read by key | Hash Map | Hash Map with Perfect Hashing | Avoid hash collisions |
+| Get documents by keyword | Inverted Index | Inverted Index with Sorted </br> and BitPacked Postings | Less memory usage, faster search |
+| Vector Search | HNSW graph | HNSW graph with </br> paylaod-aware connections | Better precision with filters |
+| Tenant isolation | Vector Storage | Defragmented Vector Storage | Faster access to on-disk data |
+
+About payload-aware connections in HNSW you can read in our [previous article](/articles/filtrable-hnsw/).
+
+But in this article I would like to describe a few of the latest additions to Qdrant: immutable hash map with perfect hashing and defragmented vector storage.
+
+### Perfect Hashing
+
+ToDo: Here describe how hash table with perfect hashing works and why it is important (see notion for details)
+
+### Deftagmentation
+
+ToDo
+
+* Describe why and when it is useful
+  * Disk is accessed by pages
+  * If page is bigger than vector, we waste cache
+  * If all relevant vectors for the tenant are together, we don't have cache misses even if vectors are small
+
+(Use image with defragmentation here)
+
+
+## How to update immutable data structures?
+
+ToDo.
+
+Basically, two ideas:
+
+- Copy-on-write
+- Soft-delete
+
+Also might we interesting to describe segments and how we can write to the segment, which is currently under optimization (and have to be read-only).
```

Comment:
```suggestion
Also, it might be interesting to describe segments and how we can write to the segment, which is currently under optimization (and have to be read-only).
```